# Reward luxembourgish instruction tuning set

In this notebook, we reward a synthetically created luxembourgish instruction tuning set, which was created with Llama-3.1-405b-instruct model from luxembourgish Wikipedia and RTL news data. The original dataset contains 2420 rows, 1930 of which were generated from RTL data and 490 from Wikipedia data. We decided however to only evaluate a subset of 200 randomly sampled entries for helpfulness, correctnes and coherence, using Nvidia-Nemotron-340b-reward. We will sample 80% from the RTL data and 20% from the wiki data, keeping approximately the distribution in the original dataset. Since the model only learned to evaluate English language data, we translated the dataset from Luxembourgish into English using Google translate.

In [ ]:
# Mount Google Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# read in the english csv file
import pandas as pd
df = pd.read_csv('/content/drive/My Drive/Master Thesis/instruction_tuning_set_en.csv')
df

instruction  \
0     Give the most important information about Debo...   
1     What is the most famous performance of Deborah...   
2     What is the background of the performance of D...   
3     What was Deborah De Robertis doing at the Cent...   
4     How was Deborah De Robertis charged in court o...   
...                                                 ...   
2415  What was the context of the national debt in t...   
2416       How much should the US debt limit be raised?   
2417           What are the reactions to the agreement?   
2418  What are the consequences if the agreement is ...   
2419  How did the stock markets react to the agreement?   

                                               response  \
0     Deborah De Robertis is a Luxembourgish-French ...   
1     The most famous performance of Deborah De Robe...   
2     On August 31, 2018, Deborah De Robertis posed ...   
3     On May 6, 2024, Deborah De Robertis hung up an...   
4     On 29 May 2024, Deborah De Robertis was charge...   
...                                                 ...   
2415  On August 1, 2011, President Barack Obama anno...   
2416  The debt limit should be increased in two stag...   
2417  Not everyone agrees with this compromise. Resi...   
2418  If the agreement is not accepted, the United S...   
2419  Stock markets around the world reacted positiv...   

                                                   text  
0     The following provides an instruction in the f...  
1     The following provides an instruction in the f...  
2     The following provides an instruction in the f...  
3     The following provides an instruction in the f...  
4     The following provides an instruction in the f...  
...                                                 ...  
2415  The following provides an instruction in the f...  
2416  The following provides an instruction in the f...  
2417  The following provides an instruction in the f...  
2418  The following provides an instruction in the f...  
2419  The following provides an instruction in the f...  

[2420 rows x 3 columns]

# Sample Data

We split the dataset into two parts. The first part contains all rows that were generated from the wikipedia data, rows 0 to 489 and the second part, contaning all rows that were generated from the RTL data, rows 490 to 2419. We then randomly sample 40 rows from the wiki data and 160 rows from the RTL data, maintaining the 0.2/0.8 split from the original dataset.

In [ ]:
wiki_df = df.iloc[:490]
rtl_df = df.iloc[490:]
print(len(wiki_df))
print(len(rtl_df))

490
1930


In [ ]:
# random sample 40 rows from wiki data
sampled_wiki_df = wiki_df.sample(n=40, random_state=42)
sampled_wiki_df = sampled_wiki_df.reset_index(drop=True) # reset index and drop orignal index
sampled_wiki_df

instruction  \
0               What is the brightest star in Dorado?   
1   What happened to the Edict of Nantes, after He...   
2   When did the devolution war start and how did ...   
3                  When and where did Max Ophüls die?   
4   What happened to the Schumacher building after...   
5   What are the essential amino acids? List the 9...   
6   Give me the names of the former municipalities...   
7   Give me the area of the Polvermillen and what ...   
8   When did Victor Tedesco start his prison sente...   
9   How many films does the Cinémathèque de la Vil...   
10  How was Leo VIII. for the second time to the P...   
11  What kind of competitions does the Music Acade...   
12  What was Martin Schwarzschild's main area of w...   
13                   When did Mathias Thinnes retire?   
14  Give the history of the cathedral Notre-Dame o...   
15  Give me the name of the Luxembourgish doom met...   
16  What was the devolution war and how did it pla...   
17  What were the goals of King Louis XIV. during ...   
18  Given the date of the dedication of the cathed...   
19  What was the legitimacy of the Hellenistic mon...   
20  When did Jean-Baptiste Kaeuffer leave his parish?   
21  When did Heinrich Rütten become the county cou...   
22  What did Martin Schwarzschild along with Lyman...   
23  Give me the name of the resistance fighter who...   
24  What was the nomination of Jean-Baptiste Kaeuf...   
25  When did Jean-Baptiste Kaeuffer swear the repu...   
26  What became of the writer's home after the Jes...   
27        Give me a short biography of Marcel Gilles.   
28  What was Jayson Sherlock's goal when he founde...   
29                  How deep is Lake Erie on average?   
30        Where was the series Dawson's Creek filmed?   
31  What is the highest point in the municipality ...   
32  Enter the list of the 20 (+ 1) amino acids tha...   
33  What is the formula for calculating the amount...   
34  What is the description of the coat of arms of...   
35       Describe the musical style of Plaguewielder.   
36  Give the most important information about Debo...   
37                           Where is Niedderkärjeng?   
38   Which company has Virgin invested in since 2017?   
39  What was the subject of the commemorative coin...   

                                             response  \
0   α Doradus, the brightest star in Dorado, is 20...   
1   The Edict of Nantes was unfortunately not "per...   
2   The war of devolution began in the spring of 1...   
3       Max Ophüls died on March 26, 1957 in Hamburg.   
4   The police museum of the non-profit organizati...   
5   The essential amino acids are: histidine, isol...   
6   The former municipalities of Beho, Bovigny, Ch...   
7   The Polvermillen has an area of 24.82 ha, whic...   
8   After the failed expedition of Exilbelsch in t...   
9   The Cinémathèque de la Ville de Luxembourg pro...   
10  Leo VIII. became pope for the second time afte...   
11  The Music Academy Franz Liszt Weimar organizes...   
12        His main area of work was star development.   
13                               On September 1, 1978   
14  The Cathedral of Notre-Dame of Luxembourg was ...   
15                                     plague wielder   
16  The War of Devolution was a military conflict ...   
17  The goals of King Louis XIV. were to expand te...   
18  On December 8, 1963, Bishop Léon Lommel solemn...   
19  The Hellenistic monarchy was personal. This me...   
20  Jean-Baptiste Kaeuffer left his parish on Febr...   
21                                   On July 5, 1944.   
22  Together with Lyman Spitzer and James Van Alle...   
23                                       Emile Benoit   
24  On August 11, 1791, Jean-Baptiste Kaeuffer was...   
25  Jean-Baptiste Kaeuffer swore the republican oa...   
26  In the writer's home, the artisan school, the ...   
27  Marcel Victor Gilles, born on June 29, 1945 in...   
28  The goal of Jayson Sherlock was to create a Ch.

In [ ]:
# now same for RTL data, but with 160 rows
sampled_rtl_df = rtl_df.sample(n=160, random_state=42)
sampled_rtl_df = sampled_rtl_df.reset_index(drop=True)
sampled_rtl_df

instruction  \
0                When did Marco Mille work at Siemens?   
1    What is the result of the large-scale police c...   
2    Give me the number of young people in Luxembou...   
3    What did the owner of an animal say about the ...   
4    Since 2022, how many agencies are responsible ...   
..                                                 ...   
155  Are there other opportunities to become politi...   
156     When and where does the City Night Bus CN1 go?   
157  What was the main topic of the consultations o...   
158  What is the role of the hereditary grand ducal...   
159             Who won the cup in basketball in 2004?   

                                              response  \
0       Marco Mille only joined Siemens in March 2010.   
1    The police issued 20 warning charges and one p...   
2    In 2019, 9.75 percent of children were fundame...   
3    The owner said that the accused did not have e...   
4    Since 2022, six agencies will be responsible f...   
..                                                 ...   
155  Yes, there are other opportunities to become p...   
156  The City Night Bus CN1 goes from the Upper Tow...   
157  The main topic was the refugee issue, especial...   
158  We really live in very complicated times and y...   
159  The Musel Pikes won the cup in basketball in 2...   

                                                  text  
0    The following provides an instruction in the f...  
1    The following provides an instruction in the f...  
2    The following provides an instruction in the f...  
3    The following provides an instruction in the f...  
4    The following provides an instruction in the f...  
..                                                 ...  
155  The following provides an instruction in the f...  
156  The following provides an instruction in the f...  
157  The following provides an instruction in the f...  
158  The following provides an instruction in the f...  
159  The following provides an instruction in the f...  

[160 rows x 3 columns]

In [ ]:
# Now we merge the dataframes back together
df = pd.concat([sampled_wiki_df, sampled_rtl_df])
df = df.reset_index(drop=True)
df

instruction  \
0                What is the brightest star in Dorado?   
1    What happened to the Edict of Nantes, after He...   
2    When did the devolution war start and how did ...   
3                   When and where did Max Ophüls die?   
4    What happened to the Schumacher building after...   
..                                                 ...   
195  Are there other opportunities to become politi...   
196     When and where does the City Night Bus CN1 go?   
197  What was the main topic of the consultations o...   
198  What is the role of the hereditary grand ducal...   
199             Who won the cup in basketball in 2004?   

                                              response  \
0    α Doradus, the brightest star in Dorado, is 20...   
1    The Edict of Nantes was unfortunately not "per...   
2    The war of devolution began in the spring of 1...   
3        Max Ophüls died on March 26, 1957 in Hamburg.   
4    The police museum of the non-profit organizati...   
..                                                 ...   
195  Yes, there are other opportunities to become p...   
196  The City Night Bus CN1 goes from the Upper Tow...   
197  The main topic was the refugee issue, especial...   
198  We really live in very complicated times and y...   
199  The Musel Pikes won the cup in basketball in 2...   

                                                  text  
0    The following provides an instruction in the f...  
1    The following provides an instruction in the f...  
2    The following provides an instruction in the f...  
3    The following provides an instruction in the f...  
4    The following provides an instruction in the f...  
..                                                 ...  
195  The following provides an instruction in the f...  
196  The following provides an instruction in the f...  
197  The following provides an instruction in the f...  
198  The following provides an instruction in the f...  
199  The following provides an instruction in the f...  

[200 rows x 3 columns]

In [ ]:
# check the prompt for correct format
for i in range(len(df[:5])):
  prompt = [{"role":"user","content":df["instruction"][i]},{"role":"assistant","content":df["response"][i]}]
  print(prompt)

[{'role': 'user', 'content': 'What is the brightest star in Dorado?'}, {'role': 'assistant', 'content': 'α Doradus, the brightest star in Dorado, is 200 light years away from us. The star is slightly white and belongs to the spectral class A0 III.'}]
[{'role': 'user', 'content': 'What happened to the Edict of Nantes, after Henri IV. died?'}, {'role': 'assistant', 'content': 'The Edict of Nantes was unfortunately not "perpétuel et irrevocable", which Louis XIV. evidenced by the Edict of Fontainebleau in 1685.'}]
[{'role': 'user', 'content': 'When did the devolution war start and how did it end?'}, {'role': 'assistant', 'content': 'The war of devolution began in the spring of 1667, when Louis XIV. sent a declaration to the Spanish court, and ended on May 2, 1668, when the Peace Treaty of Oachen was signed.'}]
[{'role': 'user', 'content': 'When and where did Max Ophüls die?'}, {'role': 'assistant', 'content': 'Max Ophüls died on March 26, 1957 in Hamburg.'}]
[{'role': 'user', 'content': '

# NIM PART - PROCEED WITH CAUTION !!!!!!!!!!!!!!!!!!!

In [ ]:
# !pip install -q openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 365.7/365.7 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 318.9/318.9 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 2.5 MB/s eta 0:00:00


In [ ]:
# from openai import OpenAI
# from google.colab import userdata

In [ ]:
# client = OpenAI(
#   base_url = "https://integrate.api.nvidia.com/v1",
#   api_key = userdata.get("NVIDIA_NIM")
# )

In [ ]:
# model_responses = []

In [ ]:
# for i in range(len(df)):

#   completion = client.chat.completions.create(
#     model="nvidia/nemotron-4-340b-reward",
#     messages=[{"role":"user","content":df["instruction"][i]},{"role":"assistant","content":df["response"][i]}],
#   )

#   # The nemotron model returns messages as a single element list containing
#   # ChatCompletionMessage, hence access with message[0].content instead of
#   # message.content like done in instruction models
#   answer = completion.choices[0].message[0].content
#   model_responses.append(answer)  # store model answer

In [ ]:
# model_responses

['helpfulness:1.84375,correctness:1.6640625,coherence:3.546875,complexity:1.0078125,verbosity:0.73046875',
 'helpfulness:2.75,correctness:2.890625,coherence:3.765625,complexity:1.1171875,verbosity:0.4765625',
 'helpfulness:2.578125,correctness:2.75,coherence:3.6875,complexity:1.1171875,verbosity:0.75390625',
 'helpfulness:3.171875,correctness:3.21875,coherence:3.984375,complexity:0.66796875,verbosity:0.484375',
 'helpfulness:2.890625,correctness:3.078125,coherence:3.75,complexity:1.140625,verbosity:1.1640625',
 'helpfulness:3.46875,correctness:3.5625,coherence:3.859375,complexity:1.078125,verbosity:0.76953125',
 'helpfulness:3.78125,correctness:3.796875,coherence:4.0,complexity:1.15625,verbosity:0.984375',
 'helpfulness:2.328125,correctness:2.515625,coherence:3.84375,complexity:1.15625,verbosity:0.81640625',
 'helpfulness:0.02294921875,correctness:-0.1982421875,coherence:2.78125,complexity:0.310546875,verbosity:0.27734375',
 'helpfulness:2.890625,correctness:2.859375,coherence:3.9375,c

# Save responses

In [ ]:
# model_responses

['helpfulness:1.84375,correctness:1.6640625,coherence:3.546875,complexity:1.0078125,verbosity:0.73046875',
 'helpfulness:2.75,correctness:2.890625,coherence:3.765625,complexity:1.1171875,verbosity:0.4765625',
 'helpfulness:2.578125,correctness:2.75,coherence:3.6875,complexity:1.1171875,verbosity:0.75390625',
 'helpfulness:3.171875,correctness:3.21875,coherence:3.984375,complexity:0.66796875,verbosity:0.484375',
 'helpfulness:2.890625,correctness:3.078125,coherence:3.75,complexity:1.140625,verbosity:1.1640625',
 'helpfulness:3.46875,correctness:3.5625,coherence:3.859375,complexity:1.078125,verbosity:0.76953125',
 'helpfulness:3.78125,correctness:3.796875,coherence:4.0,complexity:1.15625,verbosity:0.984375',
 'helpfulness:2.328125,correctness:2.515625,coherence:3.84375,complexity:1.15625,verbosity:0.81640625',
 'helpfulness:0.02294921875,correctness:-0.1982421875,coherence:2.78125,complexity:0.310546875,verbosity:0.27734375',
 'helpfulness:2.890625,correctness:2.859375,coherence:3.9375,c

In [ ]:
# df['reward'] = model_responses
# df

instruction  \
0                What is the brightest star in Dorado?   
1    What happened to the Edict of Nantes, after He...   
2    When did the devolution war start and how did ...   
3                   When and where did Max Ophüls die?   
4    What happened to the Schumacher building after...   
..                                                 ...   
195  Are there other opportunities to become politi...   
196     When and where does the City Night Bus CN1 go?   
197  What was the main topic of the consultations o...   
198  What is the role of the hereditary grand ducal...   
199             Who won the cup in basketball in 2004?   

                                              response  \
0    α Doradus, the brightest star in Dorado, is 20...   
1    The Edict of Nantes was unfortunately not "per...   
2    The war of devolution began in the spring of 1...   
3        Max Ophüls died on March 26, 1957 in Hamburg.   
4    The police museum of the non-profit organizati...   
..                                                 ...   
195  Yes, there are other opportunities to become p...   
196  The City Night Bus CN1 goes from the Upper Tow...   
197  The main topic was the refugee issue, especial...   
198  We really live in very complicated times and y...   
199  The Musel Pikes won the cup in basketball in 2...   

                                                  text  \
0    The following provides an instruction in the f...   
1    The following provides an instruction in the f...   
2    The following provides an instruction in the f...   
3    The following provides an instruction in the f...   
4    The following provides an instruction in the f...   
..                                                 ...   
195  The following provides an instruction in the f...   
196  The following provides an instruction in the f...   
197  The following provides an instruction in the f...   
198  The following provides an instruction in the f...   
199  The following provides an instruction in the f...   

                                        model_response  
0    helpfulness:1.84375,correctness:1.6640625,cohe...  
1    helpfulness:2.75,correctness:2.890625,coherenc...  
2    helpfulness:2.578125,correctness:2.75,coherenc...  
3    helpfulness:3.171875,correctness:3.21875,coher...  
4    helpfulness:2.890625,correctness:3.078125,cohe...  
..                                                 ...  
195  helpfulness:2.328125,correctness:2.53125,coher...  
196  helpfulness:0.94140625,correctness:1.0390625,c...  
197  helpfulness:2.671875,correctness:2.671875,cohe...  
198  helpfulness:1.171875,correctness:1.2734375,coh...  
199  helpfulness:0.9921875,correctness:0.83984375,c...  

[200 rows x 4 columns]

In [ ]:
# df.to_csv('/content/drive/My Drive/Master Thesis/sampled_instruction_tuning_set_reward.csv', index=False)

# Import Data as Dataframe

In [ ]:
# # read in the csv file
# import pandas as pd
# df = pd.read_csv('/content/drive/My Drive/Master Thesis/sampled_instruction_tuning_set_reward.csv')
# df

instruction  \
0                What is the brightest star in Dorado?   
1    What happened to the Edict of Nantes, after He...   
2    When did the devolution war start and how did ...   
3                   When and where did Max Ophüls die?   
4    What happened to the Schumacher building after...   
..                                                 ...   
195  Are there other opportunities to become politi...   
196     When and where does the City Night Bus CN1 go?   
197  What was the main topic of the consultations o...   
198  What is the role of the hereditary grand ducal...   
199             Who won the cup in basketball in 2004?   

                                              response  \
0    α Doradus, the brightest star in Dorado, is 20...   
1    The Edict of Nantes was unfortunately not "per...   
2    The war of devolution began in the spring of 1...   
3        Max Ophüls died on March 26, 1957 in Hamburg.   
4    The police museum of the non-profit organizati...   
..                                                 ...   
195  Yes, there are other opportunities to become p...   
196  The City Night Bus CN1 goes from the Upper Tow...   
197  The main topic was the refugee issue, especial...   
198  We really live in very complicated times and y...   
199  The Musel Pikes won the cup in basketball in 2...   

                                                  text  \
0    The following provides an instruction in the f...   
1    The following provides an instruction in the f...   
2    The following provides an instruction in the f...   
3    The following provides an instruction in the f...   
4    The following provides an instruction in the f...   
..                                                 ...   
195  The following provides an instruction in the f...   
196  The following provides an instruction in the f...   
197  The following provides an instruction in the f...   
198  The following provides an instruction in the f...   
199  The following provides an instruction in the f...   

                                        model_response  
0    helpfulness:1.84375,correctness:1.6640625,cohe...  
1    helpfulness:2.75,correctness:2.890625,coherenc...  
2    helpfulness:2.578125,correctness:2.75,coherenc...  
3    helpfulness:3.171875,correctness:3.21875,coher...  
4    helpfulness:2.890625,correctness:3.078125,cohe...  
..                                                 ...  
195  helpfulness:2.328125,correctness:2.53125,coher...  
196  helpfulness:0.94140625,correctness:1.0390625,c...  
197  helpfulness:2.671875,correctness:2.671875,cohe...  
198  helpfulness:1.171875,correctness:1.2734375,coh...  
199  helpfulness:0.9921875,correctness:0.83984375,c...  

[200 rows x 4 columns]